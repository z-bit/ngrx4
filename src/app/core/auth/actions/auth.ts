import { Action } from '@ngrx/store';
export interface AuthAction extends Action {
  type: string;
  payload?: any;
}
import { User } from '../models/user';
import { Authenticate } from '../models/authenticate';

export const LOGIN = '[Auth] Login';
export const LOGIN_SUCCESS = '[Auth] LOgin Success';
export const LOGIN_FAILURE = '[Auth] Login Failure';
export const LOGIN_REDIRECT = '[Auth] Login Redirect';
export const LOGOUT = '[Auth] Logout';

export class Login implements AuthAction {
  readonly type = LOGIN;
  constructor(
    public payload: Authenticate
  ) {}
}

export class LoginSuccess implements AuthAction {
  readonly type = LOGIN_SUCCESS;
  constructor(
    public payload: { user: User }
  ) {}
}

export class LoginFailure implements AuthAction {
  readonly type = LOGIN_FAILURE;
  constructor(
    public payload: any
  ) {}
}

export class LoginRedirect implements AuthAction {
  readonly type = LOGIN_REDIRECT;
}

export class Logout implements AuthAction {
  readonly type = LOGOUT;
}

export type Actions
  = Login
  | LoginSuccess
  | LoginFailure
  | LoginRedirect
  | Logout
;
