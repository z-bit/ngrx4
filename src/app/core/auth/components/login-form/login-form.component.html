<md-card>
  <md-card-title>Login</md-card-title>
  <md-card-content>
    <form [formGroup]="form" (ngSubmit)="submit()">
      <p>
        <md-input-container>
          <input type="text" mdInput placeholder="Username" formControlName="username">
        </md-input-container>
      </p>

      <p>
        <md-input-container>
          <input type="password" mdInput placeholder="Password" formControlName="password">
        </md-input-container>
      </p>

      <p *ngIf="errorMessage" class="loginError">
        {{ errorMessage }}
      </p>

      <p class="loginButtons">
        <button type="submit" md-button>Login</button>
      </p>

    </form>
  </md-card-content>
</md-card>