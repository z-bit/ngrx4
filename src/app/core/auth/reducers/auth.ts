import * as authActions from '../actions/auth';
import { User } from '../models/user';

export interface State {
  loggedIn: boolean;
  user: User | null;
}

export const initialState: State = {
  loggedIn: false,
  user: null,
};

export function reducer(state = initialState, action: authActions.Actions) {
  switch (action.type) {
    case authActions.LOGIN_SUCCESS: {
      return {
        ...state,
        loggedIn: true,
        user: action.payload.user,
      };
    }
    case authActions.LOGOUT: {
      return initialState;
    }
    default: {
      return state;
    }
  }
}

export const getLoggedIn = (state: State) => state.loggedIn;
export const getUser = (state: State) => state.user;
