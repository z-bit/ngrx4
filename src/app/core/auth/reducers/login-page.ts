import * as authActions from '../actions/auth';

export interface State  {
  error: string | null;
  pending: boolean;
}

export const initialState = {
  error: null,
  pending: false,
};

export function reducer(state = initialState, action: authActions.Actions) {
  switch (action.type) {
    case authActions.LOGIN: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }
    case authActions.LOGIN_SUCCESS: {
      return {
        ...state,
        error: null,
        pending: false,
      };
    }
    case authActions.LOGIN_FAILURE: {
      return {
        ...state,
        error: action.payload,
        pending: false,
      };
    }
    default: {
      return state;
    }
  }
}

export const getError = (state: State) => state.error;
export const getPending = (state: State) => state.pending;
