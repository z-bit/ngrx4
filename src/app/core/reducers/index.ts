import {
  ActionReducerMap, createSelector, createFeatureSelector,
  ActionReducer, MetaReducer } from '@ngrx/store';
import { environment } from '../../../environments/environment';
import { RouterStateUrl } from '../shared/utils';
import * as routerStore from '@ngrx/router-store';

import { storeFreeze } from 'ngrx-store-freeze';

import * as layoutReducers from '../ngrx-core/reducers/layout';

export interface State {
  layout: layoutReducers.State;
  routerReducer: routerStore.RouterReducerState<RouterStateUrl>;
}

export const reducers: ActionReducerMap<State> = {
  layout: layoutReducers.reducer,
  routerReducer: routerStore.routerReducer,
};

// console.log all actions
export function logger(reducer: ActionReducer<State>): ActionReducer<State> {
  return function(state: State, action: any): State {
    console.log('state', state);
    console.log('action', action);
    return reducer(state, action);
  };
}

export const metaReducers: MetaReducer<State>[]
  = !environment.production
  ? [logger, storeFreeze]
  : []
;

export const getLayoutState = createFeatureSelector<layoutReducers.State>('layout');
export const getShowSidenav = createSelector(
  getLayoutState,
  layoutReducers.getShowSidenav
);
