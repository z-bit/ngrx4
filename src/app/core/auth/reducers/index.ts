import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as rootReducers from '../../reducers';
import * as authReducers from './auth';
import * as loginPageReducers from './login-page';

export interface AuthState {
  status: authReducers.State;
  loginPage: loginPageReducers.State;
}

export interface State extends rootReducers.State {
  auth: AuthState;
}

export const reducers = {
  status: authReducers.reducer,
  loginPage: loginPageReducers.reducer,
};

export const selectAuthState = createFeatureSelector<AuthState>('auth');

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state: AuthState) => state.status
);
export const getLoggedIn = createSelector(
  selectAuthStatusState,
  authReducers.getLoggedIn
);
export const getUser = createSelector(
  selectAuthStatusState,
  authReducers.getUser
);
export const selectLoginPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginPage
);
export const getLoginPageError = createSelector(
  selectLoginPageState,
  loginPageReducers.getError
);
export const getLoginPagePending = createSelector(
  selectLoginPageState,
  loginPageReducers.getPending
);

