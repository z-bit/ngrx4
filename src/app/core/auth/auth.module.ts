import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LoginFormComponent } from './components/login-form/login-form.component';


@NgModule({
  imports: [
    CommonModule
  ],
  declarations: [
    LoginFormComponent,
  ]
})
export class AuthModule { }
