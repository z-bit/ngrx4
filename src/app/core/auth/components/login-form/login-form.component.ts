import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { Authenticate } from '../../models/authenticate';

@Component({
  selector: 'app-login-form',
  styleUrls: ['./login-form.component.css'],
  templateUrl: './login-form.component.html',
})
export class LoginFormComponent {
  @Input()
  set pending(isPending: boolean) {
    if (isPending) {
      this.form.disable();
    } else {
      this.form.enable();
    }
  }

  @Input() errorMessage: string | null;

  @Input() submitted = new EventEmitter<Authenticate>();

  form: FormGroup = new FormGroup({
    username: new FormControl(''),
    password: new FormControl(''),
  });

  submit() {
    if (this.form.valid) {
      this.submitted.emit(this.form.value);
    }
  }
}
