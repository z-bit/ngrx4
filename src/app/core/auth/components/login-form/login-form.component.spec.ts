import { TestBed, ComponentFixture } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { StoreModule, Store, combineReducers } from '@ngrx/store';
import { LoginFormComponent } from './login-form.component';
import * as authActions from '../../actions/auth';
import * as authReducers from '../../reducers';
import { ReactiveFormsModule } from '@angular/forms';

describe('Login-Page', () => {
  let fixture: ComponentFixture<LoginFormComponent>;
  let instance: LoginFormComponent;

  beforeEach( () => {
    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule],
      declarations: [LoginFormComponent],
      schemas: [NO_ERRORS_SCHEMA],
    });
    fixture = TestBed.createComponent(LoginFormComponent);
    instance = fixture.componentInstance;
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should disable the form if pending', () => {
    instance.pending = true;
    expect(fixture).toMatchSnapshot();
  });

  it('should display an error message if provided', () => {
    instance.errorMessage = 'Invalid credentials';
    expect(fixture).toMatchSnapshot();
  });

  it('should emit event if form is valid when submitted', () => {
    const credentials = {
      username: 'user',
      password: 'pass',
    };
    instance.form.setValue(credentials);
    spyOn(instance.submitted, 'emit');
    instance.submit();
    expect(instance.submitted.emit).toHaveBeenCalledWith(credentials);
  });
});
