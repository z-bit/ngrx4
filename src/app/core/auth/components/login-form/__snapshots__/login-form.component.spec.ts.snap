// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login-Page should compile 1`] = `
<app-login-form
  form={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <md-card>
    
    
    <md-card-title>
      Login
    </md-card-title>
    
    
    <md-card-content>
      
      
      <form
        class="ng-untouched ng-pristine ng-valid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="username"
              mdinput=""
              ng-reflect-name="username"
              placeholder="Username"
              type="text"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="password"
              mdinput=""
              ng-reflect-name="password"
              placeholder="Password"
              type="password"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        
        
  
        
        <p
          class="loginButtons"
        >
          
          
          <button
            md-button=""
            type="submit"
          >
            Login
          </button>
          
        
        </p>
        
  
      
      </form>
      
    
    </md-card-content>
    
  
  </md-card>
</app-login-form>
`;

exports[`Login-Page should disable the form if pending 1`] = `
<app-login-form
  form={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <md-card>
    
    
    <md-card-title>
      Login
    </md-card-title>
    
    
    <md-card-content>
      
      
      <form
        novalidate=""
      >
        
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              formcontrolname="username"
              mdinput=""
              placeholder="Username"
              type="text"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              formcontrolname="password"
              mdinput=""
              placeholder="Password"
              type="password"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        
        
  
        
        <p
          class="loginButtons"
        >
          
          
          <button
            md-button=""
            type="submit"
          >
            Login
          </button>
          
        
        </p>
        
  
      
      </form>
      
    
    </md-card-content>
    
  
  </md-card>
</app-login-form>
`;

exports[`Login-Page should display an error message if provided 1`] = `
<app-login-form
  errorMessage={[Function String]}
  form={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <md-card>
    
    
    <md-card-title>
      Login
    </md-card-title>
    
    
    <md-card-content>
      
      
      <form
        novalidate=""
      >
        
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              formcontrolname="username"
              mdinput=""
              placeholder="Username"
              type="text"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        <p>
          
          
          <md-input-container>
            
            
            <input
              formcontrolname="password"
              mdinput=""
              placeholder="Password"
              type="password"
            />
            
          
          </md-input-container>
          
        
        </p>
        
  
        
        
        
  
        
        <p
          class="loginButtons"
        >
          
          
          <button
            md-button=""
            type="submit"
          >
            Login
          </button>
          
        
        </p>
        
  
      
      </form>
      
    
    </md-card-content>
    
  
  </md-card>
</app-login-form>
`;
